model/modeldecoder/generator/validation.go

```diff
@@ -0,0 +1,245 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package generator
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+const (
+	tagEnum          = "enum"
+	tagMax           = "max"
+	tagMaxVals       = "maxVals"
+	tagMin           = "min"
+	tagPattern       = "pattern"
+	tagPatternKeys   = "patternKeys"
+	tagRequired      = "required"
+	tagRequiredOneOf = "requiredOneOf"
+	tagRequiredIfAny = "requiredIfAny"
+	tagTypes         = "types"
+	tagTypesVals     = "typesVals"
+)
+
+// vFieldFn should be used for implementing field specific validation rules,
+// independend of other fields or tags
+type vFieldFn func(io.Writer, structField, validationRule) error
+
+// vFieldTagsFn should be used for implementing field specific validation rules,
+// dependend on other tags, but independend of other fields
+type vFieldTagsFn func(io.Writer, structField, []validationRule, validationRule) error
+
+// vStructFn should be used for implementing struct specific validation rules,
+// depending on multiplefields
+type vStructFn func(io.Writer, []structField, structField, validationRule) error
+
+type validationFunctions struct {
+	vFieldFns     map[string]vFieldFn
+	vFieldTagsFns map[string]vFieldTagsFn
+	vStructFns    map[string]vStructFn
+}
+
+type validationRule struct {
+	name  string
+	value string
+}
+
+func validation(validationFns validationFunctions, w io.Writer, fields []structField, f structField) error {
+	vTag, err := validationTag(f.tag)
+	if err != nil {
+		return err
+	}
+	rules := validationRules(vTag)
+	for _, rule := range rules {
+		if validationFns.vFieldFns != nil {
+			if fn, ok := validationFns.vFieldFns[rule.name]; ok {
+				if err := fn(w, f, rule); err != nil {
+					return err
+				}
+				continue
+			}
+		}
+		if validationFns.vFieldTagsFns != nil {
+			if fn, ok := validationFns.vFieldTagsFns[rule.name]; ok {
+				if err := fn(w, f, rules, rule); err != nil {
+					return err
+				}
+				continue
+			}
+		}
+		if validationFns.vStructFns != nil {
+			if fn, ok := validationFns.vStructFns[rule.name]; ok {
+				if err := fn(w, fields, f, rule); err != nil {
+					return err
+				}
+				continue
+			}
+		}
+		return fmt.Errorf("unhandled tag rule '%s'", rule)
+	}
+	return nil
+}
```

Comment:
I also wonder if we need this, or if we could instead switch on the rule names internally in the generators. The latter would seem simpler to me. Perhaps later?

Reply:
changed it :)